### DISCUSSION TITLE
Global setTimeout

### DISCUSSION THREAD

**Discussion by rMinion** [jspsych/jsPsych#1859] [Q&A] [answered]
I want to create a timer that operates across different sets of trials. I want to define it on the procedure timeline level.  After starting the procedure it should choose a random time after which it changes the size and colour of certain span element, waits for 1s, and after that return parameres to default. I thought that I could achieve this with nested setTimeout but unfortunately, I've encountered some problems.  In my particular case pressing a key ends up the trial and this blocks somehow the execution of setTimeout callback.  
If the participant does not press any key the code works as intended. Is there a way to set up a global timeout within jsPsych that won't be affected by ending the trial? 

Minimal valid example in the attachment. 
[example.txt](https://github.com/jspsych/jsPsych/files/6545301/example.txt)

**Comment by becky-gilbert** [comment, marked as answer]
Hi @rMinion, the reason your `setTimeout` callback isn't running after a key press is because you're using the `jsPsych.pluginAPI.setTimeout` function, which registers the timer in the core jsPsych code so that it can be easily cleared at the end of a trial with `jsPsych.pluginAPI.clearAllTimeouts`. The html-keyboard-response plugin happens to call `jsPsych.pluginAPI.clearAllTimeouts` when the trial ends, which is also clearing your global timer ([line 83](https://github.com/jspsych/jsPsych/blob/e917fabb5541cf13800cbadd7b9b4c5a34bcd060/plugins/jspsych-html-keyboard-response.js#L83)). Using the `jsPsych.plulginAPI.setTimeout` function is generally recommended so that it can be easily cleared, because timers should almost always be cleared at the end of a trial (otherwise it will continue running during the next trial, which is typically not what we want to happen!). But of course your case is unusual in that you do want the timer to continue running throughout multiple trials. 

The solution is just to replace your `jsPsych.pluginAPI.setTimeout` function with the normal JS `setTimeout` function, so that your timer will not be registered with the jsPsych core code and therefore not be cleared by any plugins that use `jsPsych.pluginAPI.clearAllTimeouts`.

But there are a few other things I noticed, based on what I _think_ you want to do. First, if you want the red text to be shown for the duration of `red_interval`, then you actually want to immediately set the color to be red and then set up a timer to change it to black after the `red_interval` duration has passed (whereas you were calling the `red` function after `red_interval`). Similarly, if you want the text to be black for the `black_interval` duration, then you want to change the color to black and set up a timer that calls the `red` function after the `black_interval` duration.

Second, you were using `document.getElementById` to get the `keypress` element and change its color. This method usually works because normally we're changing the properties of elements during just one trial, rather than across multiple trials. But in your case, I'm guessing you want the change to the CSS properties to persist between trials, so that one one trial ends and the next one starts, the text on the next trial is still red/black if the timer hasn't timed out yet. If you use `document.getElementById`, then the change will only last for the duration of the trial. So instead I recommend creating a `style` element with a specific ID, and then accessing this element and changing the CSS rules using the `textContent` property.

Third, the `on_load` function is called at the start of _every_ trial in your procedure, and the `on_timeline_start` function is called at the start of _every_ repetition (i.e. also every trial). So this isn't a good place to initialize your first timer. Instead, I suggest using the `call-function` plugin to start the first timer right before your procedure starts. Similarly, it's good practice to clear any remaining timers at the end of the procedure, so I suggest adding a call-function trial for that as well.

So here's my solution (and apologies if this isn't what you're trying to do!)
```html
<!DOCTYPE html>
<html>
<head>
    <script src="jspsych-6.3.0/jspsych.js"></script>
    <script src="jspsych-6.3.0/plugins/jspsych-html-keyboard-response.js"></script>
    <script src="jspsych-6.3.0/plugins/jspsych-instructions.js"></script>
    <script src="jspsych-6.3.0/plugins/jspsych-call-function.js"></script>
    <link href="jspsych-6.3.0/css/jspsych.css" rel="stylesheet">
</head>
<style id="keypress-style">
    #keypress {color: red; font-size:300%;}
</style>
<body>
    <script>
        //Function to generate random numbers between min and max
        function getRandomIntInclusive(min, max) {
            min = Math.ceil(min);
            max = Math.floor(max);
            return Math.floor(Math.random() * (max - min + 1)) + min;
        }

        // change to red, and set up a timer to change to black
        function red() {
            // access the keypress-style element and change the CSS to red
            document.getElementById('keypress-style').textContent = '#keypress {color: red; font-size:300%;}';
            //setting a new random red_interval
            red_interval = getRandomIntInclusive(1, 5) * 1000
            console.log("red", red_interval)
            // change to black after the red interval ends
            red_interval_timeout = setTimeout(black, red_interval);
        }

        // change to black, and set up a timer to change to red
        function black() {
            console.log("black", black_interval)
            // access the keypress-style element and change the CSS to black
            document.getElementById('keypress-style').textContent = '#keypress {color: black; font-size:100%;}';
            // change back to red after the black interval ends
            red_interval_timeout = setTimeout(red, black_interval)
        }

        // Value in ms after which the span element should be coloured red 
        var red_interval = null

        // Value in ms after which the span element should be coloured white 
        var black_interval = 1000

        // Keypress counter
        var keypress = 0

        var start_first_timer = {
            type: 'call-function',
            func: function() {
                //setting the first value of red_interval
                red_interval = getRandomIntInclusive(1, 5) * 1000
                // change color to red
                red();
            }
        }

        // Triall 
        var letter =
        {
            type: 'html-keyboard-response',
            stimulus: ``,
            prompt: "Press a",
            choices: ["a"],
            on_start: function (letter) {
                letter.stimulus = `<span id = "keypress">${keypress}</span>`
            },
            on_finish: function () {
                keypress++
            }
        }

        var clear_last_timer = {
            type: 'call-function',
            func: function() {
                clearTimeout(red_interval_timeout);
            }
        }

        // Timeline
        var procedura = {
            timeline: [letter],
            repetitions: 100
        }

        jsPsych.init({
            timeline: [start_first_timer, procedura, clear_last_timer],
            on_finish: function () {
                jsPsych.data.displayData()
            }
        });
    </script>
</body>
</html>
```

**Reply by rMinion** [reply]
@becky-gilbert  Thank you very much! This is exactly what I was looking for. You were spot on!
